content/300-guides/050-database/052-extend-client.mdx

```diff
@@ -0,0 +1,445 @@
+      firstName: 'Alice',
+      lastName: 'Wonderland',
+      email: 'alice@prisma.io',
+      password: 'everything is awesome',
+    },
+  })
+}
+```
+
+</tab>
+
+</TabbedContent>
+
+## Validating User Input
+
+Some data must be validated at runtime. For example, there's no way to type-check that your user's email contains an `@` symbol at build time. To ensure that the data you get is the data you expect, you can use runtime validation.
+
+The Node.js ecosystem provides a number of high-quality, easy-to-use validation libraries you can choose from including: [joi](https://github.com/sideway/joi), [validator.js,](https://github.com/validatorjs/validator.js) [yup](https://github.com/jquense/yup), [zod](https://github.com/colinhacks/zod) and [superstruct](https://github.com/ianstormtaylor/superstruct).
+
+Let's go with superstruct for the example below and extend the signup helper function to ensure that the data you get meets your expectation.
+
+### Adding Validation to Your Signup Function
+
+```tsx file=index.ts highlight=2,3,7-20,29-31;normal
+import { PrismaClient, Prisma, User, Team } from '@prisma/client'
+import { assert, object, string, size, refine } from 'superstruct'
+import isEmail from 'isemail'
+
+const prisma = new PrismaClient()
+
+// Runtime validation
+const Signup = object({
+  team: object({
+    name: size(string(), 2, 30),
+  }),
+  user: object({
+    // string and an email
+    email: refine(string(), 'email', (v) => isEmail.validate(v)),
+    // password is between 7 and 30 characters long
+    password: size(string(), 7, 30),
+    firstName: size(string(), 2, 50),
+    lastName: size(string(), 2, 50),
+  }),
+})
+
+type Signup = {
+  team: Omit<Prisma.TeamCreateArgs['data'], 'id'>
+  user: Omit<Prisma.UserCreateArgs['data'], 'id'>
+}
+
+// Signup function
+async function signup(input: Signup): Promise<User> {
+  // Assert that input conforms to Signup, throwing with a helpful
+  // error message if input is invalid.
+  assert(input, Signup)
+  const team = await prisma.team.create({
+    data: {
+      ...input.team,
+      users: {
+        create: input.user,
+      },
+    },
+    include: {
+      users: true,
+    },
+  })
+  return team.users[0]
+}
+```
+
+The example aboves shows how you can easily add runtime validation to provide additionally check that the data we receive is the data we expect.
+
+## Computed Fields
+
+Another common pattern is computing a field that's derived from other data.
+
+With TypeScript's rich type system, you can define a function that accepts a generic as an input then extend that generic to ensure it conforms to a specific structure. Finally you can return that generic with additional computed fields. Let's see how that might look:
+
+Let's say we also want to derive the user's full name from their first and last name. In the example below, you'll create a helper function to compute a full name from the first and last name stored in the database.
+
+### Computing a Full Name
+
```

Comment:
How is this comment diff from what is in computed-fields.mdx? (My info architecture confusion...) Or are you doing some sort of fancy {include} and single-sourcing content?